<a href="https://colab.research.google.com/github/swapnil190301/EDA/blob/main/DA_LAB_2019120037_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**DATA ANALYTICS LAB EXPERIMENT 4: Linear Regression**

**Objective**: Building Linear Regression model for given dataset.

**Dataset**: 

Climate Change

There have been many studies documenting that the average global temperature has been increasing over the last century. The consequences of a continued rise in global temperature will be dire. Rising sea levels and an increased frequency of extreme weather events will affect billions of people.

In this problem, we will attempt to study the relationship between average global temperature and several other factors.

The file climate_change (CSV placed in lab folder) contains climate data from May 1983 to December 2008. The available variables include:

    Year: the observation year.
    Month: the observation month.
    Temp: the difference in degrees Celsius between the average global temperature in that period and a reference value. This data comes from the Climatic Research Unit at the University of East Anglia.
    CO2, N2O, CH4, CFC.11, CFC.12: atmospheric concentrations of carbon dioxide (CO2), nitrous oxide (N2O), methane  (CH4), trichlorofluoromethane (CCl3F; commonly referred to as CFC-11) and dichlorodifluoromethane (CCl2F2; commonly referred to as CFC-12), respectively. This data comes from the ESRL/NOAA Global Monitoring Division.

    CO2, N2O and CH4 are expressed in ppmv (parts per million by volume  -- i.e., 397 ppmv of CO2 means that CO2 constitutes 397 millionths of the total volume of the atmosphere)

    CFC.11 and CFC.12 are expressed in ppbv (parts per billion by volume). 

    Aerosols: the mean stratospheric aerosol optical depth at 550 nm. This variable is linked to volcanoes, as volcanic eruptions result in new particles being added to the atmosphere, which affect how much of the sun's energy is reflected back into space. This data is from the Godard Institute for Space Studies at NASA.

    TSI: the total solar irradiance (TSI) in W/m2 (the rate at which the sun's energy is deposited per unit area). Due to sunspots and other solar phenomena, the amount of energy that is given off by the sun varies substantially with time. This data is from the SOLARIS-HEPPA project website.

    MEI: multivariate El Nino Southern Oscillation index (MEI), a measure of the strength of the El Nino/La Nina-Southern Oscillation (a weather effect in the Pacific Ocean that affects global temperatures). This data comes from the ESRL/NOAA Physical Sciences Division.

In [ ]:
#importing all packages
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
sns.set_style('whitegrid')

In [ ]:
#reading the dataset
df = pd.read_csv("https://raw.githubusercontent.com/TarekDib03/Analytics/master/Week2%20-%20Linear%20Regression/Data/climate_change.csv")
df

Year  Month    MEI     CO2      CH4      N2O   CFC-11   CFC-12  \
0    1983      5  2.556  345.96  1638.59  303.677  191.324  350.113   
1    1983      6  2.167  345.52  1633.71  303.746  192.057  351.848   
2    1983      7  1.741  344.15  1633.22  303.795  192.818  353.725   
3    1983      8  1.130  342.25  1631.35  303.839  193.602  355.633   
4    1983      9  0.428  340.17  1648.40  303.901  194.392  357.465   
..    ...    ...    ...     ...      ...      ...      ...      ...   
303  2008      8 -0.266  384.15  1779.88  321.405  244.200  535.072   
304  2008      9 -0.643  383.09  1795.08  321.529  244.083  535.048   
305  2008     10 -0.780  382.99  1814.18  321.796  244.080  534.927   
306  2008     11 -0.621  384.13  1812.37  322.013  244.225  534.906   
307  2008     12 -0.666  385.56  1812.88  322.182  244.204  535.005   

           TSI  Aerosols   Temp  
0    1366.1024    0.0863  0.109  
1    1366.1208    0.0794  0.118  
2    1366.2850    0.0731  0.137  
3    1366.4202    0.0673  0.176  
4    1366.2335    0.0619  0.149  
..         ...       ...    ...  
303  1365.6570    0.0036  0.407  
304  1365.6647    0.0043  0.378  
305  1365.6759    0.0046  0.440  
306  1365.7065    0.0048  0.394  
307  1365.6926    0.0046  0.330  

[308 rows x 11 columns]

In [ ]:
#knowing the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 308 entries, 0 to 307
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Year      308 non-null    int64  
 1   Month     308 non-null    int64  
 2   MEI       308 non-null    float64
 3   CO2       308 non-null    float64
 4   CH4       308 non-null    float64
 5   N2O       308 non-null    float64
 6   CFC-11    308 non-null    float64
 7   CFC-12    308 non-null    float64
 8   TSI       308 non-null    float64
 9   Aerosols  308 non-null    float64
 10  Temp      308 non-null    float64
dtypes: float64(9), int64(2)
memory usage: 26.6 KB


In [ ]:
#correlation matrix
df.corr()

Year     Month       MEI       CO2       CH4       N2O  \
Year      1.000000 -0.025789 -0.145345  0.985379  0.910563  0.994850   
Month    -0.025789  1.000000 -0.016345 -0.096287  0.017558  0.012395   
MEI      -0.145345 -0.016345  1.000000 -0.152911 -0.105555 -0.162375   
CO2       0.985379 -0.096287 -0.152911  1.000000  0.872253  0.981135   
CH4       0.910563  0.017558 -0.105555  0.872253  1.000000  0.894409   
N2O       0.994850  0.012395 -0.162375  0.981135  0.894409  1.000000   
CFC-11    0.460965 -0.014914  0.088171  0.401284  0.713504  0.412155   
CFC-12    0.870067 -0.001084 -0.039836  0.823210  0.958237  0.839295   
TSI       0.022353 -0.032754 -0.076826  0.017867  0.146335  0.039892   
Aerosols -0.361884  0.014845  0.352351 -0.369265 -0.290381 -0.353499   
Temp      0.755731 -0.098016  0.135292  0.748505  0.699697  0.743242   

            CFC-11    CFC-12       TSI  Aerosols      Temp  
Year      0.460965  0.870067  0.022353 -0.361884  0.755731  
Month    -0.014914 -0.001084 -0.032754  0.014845 -0.098016  
MEI       0.088171 -0.039836 -0.076826  0.352351  0.135292  
CO2       0.401284  0.823210  0.017867 -0.369265  0.748505  
CH4       0.713504  0.958237  0.146335 -0.290381  0.699697  
N2O       0.412155  0.839295  0.039892 -0.353499  0.743242  
CFC-11    1.000000  0.831381  0.284629 -0.032302  0.380111  
CFC-12    0.831381  1.000000  0.189270 -0.243785  0.688944  
TSI       0.284629  0.189270  1.000000  0.083238  0.182186  
Aerosols -0.032302 -0.243785  0.083238  1.000000 -0.392069  
Temp      0.380111  0.688944  0.182186 -0.392069  1.000000

In [ ]:
#separating input and output for training the model
input = df.iloc[:,2:10]
output = df.iloc[:,10]
input

MEI     CO2      CH4      N2O   CFC-11   CFC-12        TSI  Aerosols
0    2.556  345.96  1638.59  303.677  191.324  350.113  1366.1024    0.0863
1    2.167  345.52  1633.71  303.746  192.057  351.848  1366.1208    0.0794
2    1.741  344.15  1633.22  303.795  192.818  353.725  1366.2850    0.0731
3    1.130  342.25  1631.35  303.839  193.602  355.633  1366.4202    0.0673
4    0.428  340.17  1648.40  303.901  194.392  357.465  1366.2335    0.0619
..     ...     ...      ...      ...      ...      ...        ...       ...
303 -0.266  384.15  1779.88  321.405  244.200  535.072  1365.6570    0.0036
304 -0.643  383.09  1795.08  321.529  244.083  535.048  1365.6647    0.0043
305 -0.780  382.99  1814.18  321.796  244.080  534.927  1365.6759    0.0046
306 -0.621  384.13  1812.37  322.013  244.225  534.906  1365.7065    0.0048
307 -0.666  385.56  1812.88  322.182  244.204  535.005  1365.6926    0.0046

[308 rows x 8 columns]

In [ ]:
#displaying output
output

0      0.109
1      0.118
2      0.137
3      0.176
4      0.149
       ...  
303    0.407
304    0.378
305    0.440
306    0.394
307    0.330
Name: Temp, Length: 308, dtype: float64

In [ ]:
#building model 1.1(without taking p-value into consideration)
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression 
from sklearn.metrics import r2_score

x_train,x_test,y_train,y_test = train_test_split(input,output,test_size = 0.2,random_state = 46)

In [ ]:
#fiiting the linear regression model
linreg=LinearRegression()
linreg.fit(x_train,y_train)

LinearRegression()

In [ ]:
#predicted target variable
y_pred=linreg.predict(x_test)
y_pred

array([ 0.00833482,  0.45333806,  0.32035545,  0.18549945,  0.29772647,
        0.26024605,  0.34834396,  0.00210711,  0.01328475,  0.01229301,
        0.45488459,  0.12484243,  0.39703341,  0.17171704, -0.04246809,
        0.33925564,  0.48207017,  0.45172358,  0.43504239,  0.452327  ,
        0.30964526,  0.02999622,  0.18714303,  0.13691139,  0.38455831,
        0.05505244,  0.18866127,  0.53758464,  0.21847201,  0.20260124,
        0.39868547,  0.52306189,  0.2105695 ,  0.44632212,  0.49467068,
        0.20884555,  0.36155649,  0.45418057,  0.23302812,  0.47815027,
        0.17477176,  0.02139489,  0.21407699,  0.00706016,  0.21895883,
       -0.03428431,  0.20935456,  0.35623409,  0.21057134,  0.47603081,
        0.29911427,  0.47029893,  0.49369522,  0.01354523,  0.12945259,
        0.15563692,  0.45017129,  0.02565892,  0.16210568, -0.00273264,
       -0.03981675,  0.43086001])

In [ ]:
#calculating r2-score for determining accuracy of the model
Accuracy=r2_score(y_test,y_pred)*100
print(" Accuracy of the model is %.2f" %Accuracy)

 Accuracy of the model is 72.87


In [ ]:
#checking for negative coefficients
linreg.coef_

array([ 6.69352945e-02,  4.17093430e-03,  2.76741843e-05, -1.94004437e-02,
       -8.56582542e-03,  5.17276136e-03,  9.46616546e-02, -1.55832832e+00])

In [ ]:
#calculating p-value
from scipy import stats

params = np.append(linreg.intercept_,linreg.coef_)
new_X = np.append(np.ones((len(x_train),1)), x_train, axis=1)
M_S_E = (sum((y_pred)**2))/(len(new_X)-len(new_X[0]))
v_b = M_S_E*(np.linalg.inv(np.dot(new_X.T,new_X)).diagonal())
s_b = np.sqrt(v_b)
t_b = params/ s_b
p_val =[2*(1-stats.t.cdf(np.abs(i),(len(new_X)-len(new_X[0])))) for i in t_b]
p_val = np.round(p_val,3)
p_val

array([0.001, 0.   , 0.323, 0.977, 0.193, 0.002, 0.002, 0.001, 0.   ])

In [ ]:
#dropping columns with p-value below 0.05
j=0
d=np.empty(3).astype(int)
for i in range(0,len(p_val)):
  if(p_val[i]>0.050):
    d[j]=i
    j=j+1
x_train=x_train.drop(df.columns[d],axis=1)
x_test=x_test.drop(df.columns[d],axis=1)

In [ ]:
#x_train after dropping the columns
x_train

N2O   CFC-11   CFC-12        TSI  Aerosols
7    304.082  196.609  362.174  1366.0607    0.0486
45   306.083  230.592  414.572  1365.6590    0.0111
97   308.954  267.329  492.105  1366.4313    0.0179
92   309.404  265.709  489.664  1366.0527    0.0060
137  310.813  270.928  519.545  1365.7159    0.0166
..       ...      ...      ...        ...       ...
251  318.130  254.385  542.685  1366.1742    0.0024
192  314.381  264.811  539.015  1366.3802    0.0021
117  310.344  270.814  509.287  1365.8985    0.0661
47   305.433  232.098  417.549  1365.6385    0.0104
172  312.845  267.414  534.415  1365.6734    0.0053

[246 rows x 5 columns]

In [ ]:
#x_test after dropping the columns
x_test

N2O   CFC-11   CFC-12        TSI  Aerosols
63   306.522  248.187  449.005  1366.1451    0.0096
247  318.129  255.522  543.199  1366.2899    0.0027
293  320.855  246.189  537.380  1365.7388    0.0041
124  309.942  270.685  513.455  1365.8892    0.0394
188  314.589  266.016  538.872  1366.4773    0.0022
..       ...      ...      ...        ...       ...
34   306.055  220.315  398.410  1365.6797    0.0144
108  309.383  268.650  504.226  1366.4173    0.1379
33   305.911  219.379  397.345  1365.5475    0.0158
21   305.243  209.543  374.681  1365.7085    0.0160
254  318.131  253.590  541.999  1365.7578    0.0023

[62 rows x 5 columns]

In [ ]:
#building model 1.2(taking p-value into consideration)
linreg2=LinearRegression()
linreg2.fit(x_train,y_train)

LinearRegression()

In [ ]:
#predicted target variable
y_pred=linreg2.predict(x_test)
y_pred

array([0.1165145 , 0.43287548, 0.40008182, 0.18932308, 0.38161259,
       0.25682975, 0.39020677, 0.11660333, 0.04511024, 0.12649641,
       0.41047571, 0.13468696, 0.43792134, 0.10861259, 0.01983989,
       0.41957303, 0.42545878, 0.41888015, 0.42441282, 0.42568889,
       0.38040512, 0.07467164, 0.18971988, 0.1244841 , 0.43838085,
       0.07059997, 0.20500995, 0.45139348, 0.27078356, 0.16660614,
       0.41822013, 0.45195632, 0.20776046, 0.42265319, 0.47730532,
       0.29480109, 0.39879243, 0.42620812, 0.21991224, 0.43735245,
       0.08612861, 0.14072397, 0.29523162, 0.04649355, 0.21897702,
       0.04287296, 0.21114636, 0.3062798 , 0.19168468, 0.42690363,
       0.39169039, 0.33358448, 0.44209288, 0.07051263, 0.1116168 ,
       0.10579735, 0.43054172, 0.04698035, 0.10354845, 0.04196146,
       0.01985793, 0.40919811])

In [ ]:
#calculating r2-score for determining accuracy of the model
Accuracy=r2_score(y_test,y_pred)*100
print(" Accuracy of the model is %.2f" %Accuracy)

 Accuracy of the model is 61.81


Current scientific opinion is that nitrous oxide and CFC-11 are greenhouse gases: gases that are able to trap heat from the sun and contribute to the heating of the Earth. However, the regression coefficients of both the N2O and CFC-11 variables are negative, indicating that increasing atmospheric concentrations of either of these two compounds is associated with 
lower global temperatures.

Which of the following is the simplest correct explanation for this contradiction?

I. Climate scientists are wrong that N2O and CFC-11 are greenhouse gases - this 
regression analysis constitutes part of a disproof.

II. There is not enough data, so the regression coefficients being estimated are not accurate.

III. All of the gas concentration variables reflect human development - N2O and CFC.11 are correlated with other variables in the data set.


**ANSWER:** The linear correlation of N2O and CFC.11 with other variables in the data set is quite large. The first explanation does not seem correct, as the warming effect of nitrous oxide and CFC-11 are well documented, and our regression analysis is not enough to disprove it. The second explanation is unlikely, as we have estimated eight coefficients and the intercept from 284 observations.

**OBSERVATION:**



*   The first model which was without taking p-value into consideration has accuracy of 72.87%.
*   The second model which took p-value into consideration has accuracy of 61.81%.

*   3 attributes had p-value greater than 0.05.
*   N2O and CFC-11 had negatuve coefficients of linear regression.





**INFERENCE:**



*   Decreasing no of attributes results in lowering of accuracy.
*   Linear regression isn't the most accurate model as the accuracy is around 70%.

*   I learnt to build a model and then predict the target variable.
*   I learnt about various accuracy and statistical parameters such as r2-score and p-value.



